### DISCUSSION TITLE
How to test TimeoutException for symfony/http-client

### DISCUSSION THREAD

**Discussion by homersimpsons** [symfony/symfony#58392] [Q&A] [answered]
Hello,

I have a piece of code that reacts to a `TimeoutException`, I want to test this code path. This code is also only checking for the status code to be 200. Here is a summary of my code as a test:
```php
public function testTimeout(): void
{
    $responses = [ /* A timeout response */];
    $client = new MockHttpClient($responses);

    $response = $client->request('GET', '/test');

    $this->expectException(TimeoutException::class);
    $this->assertEquals(200, $response->getStatusCode());
}
```

My issue is that I cannot seem to find a way to produce a `TimeoutException`, I always end with a `TransportException`.

I tried with those responses:
```php
$responses = [new class extends MockResponse { public function getStatusCode(): int { throw new TimeoutException(); } }];
$responses = [new MockResponse((static fn () => yield new ErrorChunk(0, 'Timeout1'))())];
$responses = [new MockResponse([new ErrorChunk(0, 'Timeout')])];
```

They all seem to give an error in their `ErrorChunk::_destruct` (or for an `ErrorChunk` that they produce, I'm not sure).

Interestingly if I remove the last assert this test pass (but in my code I cannot remove this `assert` because I have to check for the status code):
```diff
-    $this->assertEquals(200, $response->getStatusCode());
+    $response->getStatusCode();
```

Does anyone have a suggestion? Am I misunderstanding this? Is there an issue in `MockResponse` that I should report?

**Comment by HypeMC** [comment, marked as answer]
According to the [docs](https://symfony.com/doc/7.1/http_client.html#http-client-and-responses):

> However, using [MockResponse](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/HttpClient/Response/MockResponse.php) allows simulating chunked responses and timeouts:
> 
> ```php
> $body = function (): \Generator {
>     yield 'hello';
>     // empty strings are turned into timeouts so that they are easy to test
>     yield '';
>     yield 'world';
> };
> 
> $mockResponse = new MockResponse($body());
> ```
